File: backend/radio.py
"""Auto-radio: keep a room's queue topped up with related tracks.

When ``room.auto_radio`` is on and the queue nears empty, we pull the "Mix"
(radio) playlist of the last-played track from YouTube and append a handful of
fresh tracks tagged as added by 📻 Radio.
"""

import asyncio
import time
from typing import Awaitable, Callable, Optional

from config import MAX_DURATION, RADIO_BATCH, RADIO_REFILL_AT
from connections import manager
from models import Room, Track
from media_client import fetch_related, fetch_track
from media_prefetch import ensure_room_media, ensure_track_ready

RADIO_TAG = "📻 Radio"


def _seed_url(room: Room) -> str:
    """The YouTube URL whose Mix we grow the queue from."""
    if room.radio_seed_url:
        return room.radio_seed_url
    if room.queue:
        return room.queue[-1].source_url
    return ""


def needs_refill(room: Room) -> bool:
    """Whether the room is due for an auto-radio top-up right now."""
    return (
        room.auto_radio
        and not room.radio_filling
        and len(room.queue) <= RADIO_REFILL_AT
        and bool(_seed_url(room))
    )


async def _collect_tracks(room: Room, seed: str, limit: int) -> list[Track]:
    """Resolve up to ``limit`` unseen tracks from ``seed``'s YouTube Mix.

    Records every media ID it considers in ``room.radio_seen`` so a later
    refill does not offer the same track again.
    """
    candidates = await fetch_related(seed, limit * 4)

    picked: list[Track] = []
    seen_media_ids = room.radio_seen | {
        track.media_id for track in room.queue if track.media_id
    }
    for url in candidates:
        if len(picked) >= limit:
            break
        info = await fetch_track(url)
        if not info:
            continue
        media_id = info.get("media_id", "")
        if not media_id or media_id in seen_media_ids:
            continue
        duration = info.get("duration", 0) or 0
        if duration > MAX_DURATION:
            continue
        room.radio_seen.add(media_id)
        seen_media_ids.add(media_id)
        picked.append(Track.from_youtube(info, added_by=RADIO_TAG))
    return picked


async def refill(
    room: Room,
    on_start: Optional[Callable[[], Awaitable[None]]] = None,
) -> bool:
    """Append related tracks to the queue. Returns ``True`` if any were added.

    Guarded by ``room.radio_filling`` so overlapping callers don't stack
    refills on top of each other. ``on_start`` is awaited once the search is
    under way (``room.radio_filling`` set), so callers can surface a
    "searching" indicator before the network work begins.
    """
    if not needs_refill(room):
        return False

    seed = _seed_url(room)
    room.radio_filling = True
    try:
        if on_start is not None:
            await on_start()
        new_tracks = await _collect_tracks(room, seed, RADIO_BATCH)
        if not new_tracks:
            return False

        room.queue.extend(new_tracks)
        if not room.is_playing:
            room.is_playing = True
            room.position = 0.0
            room.last_sync_at = time.time()
        return True
    finally:
        room.radio_filling = False


async def refill_and_broadcast(room: Room) -> None:
    """Fire-and-forget refill that shows a "searching" state, then the result."""
    from store import save_tracks  # local import avoids an import cycle

    announced = False
    was_empty = not room.queue

    async def announce_searching() -> None:
        nonlocal announced
        announced = True
        await manager.broadcast(room.id, room.to_dict())

    added = await refill(room, on_start=announce_searching)
    if added:
        # A radio refill can start from an empty queue. Prepare its first
        # track before publishing is_playing=True to connected clients.
        if was_empty:
            room.current_index = 0
            first_track = room.queue[0] if room.queue else None
            if not await ensure_track_ready(first_track):
                room.is_playing = False
                room.position = 0.0
        await save_tracks(room)
    if added or announced:
        # Push the final state so the "searching" indicator clears, even when
        # the search turned up nothing.
        await manager.broadcast(room.id, room.to_dict())


def maybe_refill(room: Room) -> None:
    """Kick off a background refill if the room needs one.

    Cheap and safe to call after any queue/playback change; does nothing
    unless :func:`needs_refill` says so.
    """
    if needs_refill(room):
        asyncio.create_task(refill_and_broadcast(room))

File: backend/internal/room/radio.go
package room

import (
	"context"
	"time"

	"github.com/leenzstra/bebradio/backend/internal/domain"
	"github.com/leenzstra/bebradio/backend/internal/youtube"
)

// RadioTag is the "added by" credit given to tracks auto-radio queues.
const RadioTag = "\U0001F4FB Radio"

// radioTimeout bounds one refill. A batch is several yt-dlp calls in sequence,
// each of which can take tens of seconds.
const radioTimeout = 5 * time.Minute

// maybeRefill starts a background auto-radio top-up if the room is due one.
//
// It is cheap and safe to call after any change to a queue or to playback, and
// does nothing unless the room has auto-radio on, is running low, and is not
// already being refilled.
func (s *Service) maybeRefill(e *entry) {
	e.mu.Lock()
	due := s.needsRefill(e.state)
	if due {
		// Claim the refill under the same lock that decided it was needed, so
		// two callers arriving together cannot both start one.
		e.state.RadioFilling = true
	}
	e.mu.Unlock()

	if !due {
		return
	}
	s.goBackground("radio-refill", radioTimeout, func(ctx context.Context) {
		s.refill(ctx, e)
	})
}

// needsRefill reports whether a room is due an auto-radio top-up right now. The
// caller must hold the room's lock.
func (s *Service) needsRefill(r *domain.Room) bool {
	return r.AutoRadio &&
		!r.RadioFilling &&
		len(r.Queue) <= s.cfg.RadioRefillAt &&
		radioSeed(r) != ""
}

// radioSeed is the YouTube URL whose Mix the queue is grown from: the track
// that just finished, or failing that whatever is last in the queue.
func radioSeed(r *domain.Room) string {
	if r.RadioSeedURL != "" {
		return r.RadioSeedURL
	}
	if n := len(r.Queue); n > 0 {
		return r.Queue[n-1].SourceURL
	}
	return ""
}

// refill appends related tracks to a room's queue.
//
// The caller must already have claimed the refill by setting RadioFilling;
// clearing it again is this function's job, whatever happens. Listeners are
// told twice: once when the search starts, so the client can show that the room
// is looking for music, and once when it ends -- even if it found nothing, so
// the indicator always clears.
func (s *Service) refill(ctx context.Context, e *entry) {
	defer func() {
		e.mu.Lock()
		e.state.RadioFilling = false
		e.mu.Unlock()
		s.broadcast(e)
	}()

	e.mu.Lock()
	seed := radioSeed(e.state)
	seen := make(map[string]struct{}, len(e.state.RadioSeen))
	for id := range e.state.RadioSeen {
		seen[id] = struct{}{}
	}
	e.mu.Unlock()

	if seed == "" {
		return
	}
	s.broadcast(e)

	tracks := s.collectRadioTracks(ctx, e, seed, seen, s.cfg.RadioBatch)
	if len(tracks) == 0 {
		return
	}

	e.mu.Lock()
	e.state.Queue = append(e.state.Queue, tracks...)
	if !e.state.IsPlaying {
		e.state.IsPlaying = true
		e.state.Position = 0
		e.state.LastSyncAt = domain.Now()
	}
	e.mu.Unlock()

	s.persistTracks(ctx, e)
}

// collectRadioTracks resolves up to limit unseen tracks from the seed's YouTube
// Mix, recording every video id it considers so a later refill does not offer
// the same track again.
func (s *Service) collectRadioTracks(ctx context.Context, e *entry, seed string, seen map[string]struct{}, limit int) []*domain.Track {
	s.rememberRadioVideo(e, youtube.VideoID(seed))

	// Ask for more candidates than needed: some will already have been played,
	// and some will fail to resolve.
	candidates, err := s.yt.FetchRelated(ctx, seed, limit*4)
	if err != nil {
		s.log.Warn("fetching radio candidates", "seed", seed, "error", err)
		return nil
	}

	picked := make([]*domain.Track, 0, limit)
	for _, url := range candidates {
		if len(picked) >= limit || ctx.Err() != nil {
			break
		}
		vid := youtube.VideoID(url)
		if vid == "" {
			continue
		}
		if _, already := seen[vid]; already {
			continue
		}

		info, err := s.yt.FetchTrack(ctx, url)
		if err != nil {
			s.log.Debug("skipping unresolvable radio track", "url", url, "error", err)
			continue
		}

		seen[vid] = struct{}{}
		s.rememberRadioVideo(e, vid)
		picked = append(picked, domain.NewTrackFromInfo(toDomainInfo(info), RadioTag))
	}
	return picked
}

func (s *Service) rememberRadioVideo(e *entry, videoID string) {
	if videoID == "" {
		return
	}
	e.mu.Lock()
	if e.state.RadioSeen == nil {
		e.state.RadioSeen = map[string]struct{}{}
	}
	e.state.RadioSeen[videoID] = struct{}{}
	e.mu.Unlock()
}
